cli/lsp/diagnostics.rs

```diff
@@ -203,15 +131,131 @@ impl DiagnosticsServer {
+                        })
+                        .collect();
+                    }
+
+                    for (specifier, version, diagnostics) in diagnostics {
+                      client
+                        .publish_diagnostics(specifier, diagnostics, version)
+                        .await;
+                    }
+                    performance.measure(mark);
+                  }
+                }
+              }));
+
+              let previous_deps_handle = deps_handle.take();
+              deps_handle = Some(tokio::spawn({
+                let performance = performance.clone();
+                let client = client.clone();
+                let token = token.clone();
+                let snapshot = snapshot.clone();
+                let config = config.clone();
+                async move {
+                  if let Some(previous_handle) = previous_deps_handle {
+                    previous_handle.await;
+                  }
+                  let mark =
+                    performance.mark("update_diagnostics_deps", None::<()>);
+                  let diagnostics = generate_deps_diagnostics(
+                    snapshot.clone(),
+                    config.clone(),
+                    token.clone(),
+                  )
+                  .await;
+
+                  if !token.is_cancelled() {
+                    for (specifier, version, diagnostics) in diagnostics {
+                      client
+                        .publish_diagnostics(specifier, diagnostics, version)
+                        .await;
+                    }
+                    performance.measure(mark);
+                  }
+                }
+              }));
+
+              let previous_lint_handle = lint_handle.take();
+              lint_handle = Some(tokio::spawn({
+                let performance = performance.clone();
+                let client = client.clone();
+                let token = token.clone();
+                let snapshot = snapshot.clone();
+                let config = config.clone();
+                async move {
+                  if let Some(previous_handle) = previous_lint_handle {
+                    previous_handle.await;
+                  }
+                  let mark =
+                    performance.mark("update_diagnostics_lint", None::<()>);
+                  let diagnostics = generate_lint_diagnostics(
+                    &snapshot,
+                    &config,
+                    maybe_lint_config,
+                    token.clone(),
+                  )
+                  .await;
+
+                  if !token.is_cancelled() {
+                    for (specifier, version, diagnostics) in diagnostics {
+                      client
+                        .publish_diagnostics(specifier, diagnostics, version)
+                        .await;
+                    }
+                    performance.measure(mark);
+                  }
+                }
+              }));
```

Comment:
I know... this function is large. I think in a future PR it might be good to refactor this into a struct so the dependencies can be shared easily amongst the functions.